asyncio/queue.py

```diff
@@ -0,0 +1,145 @@
+# SPDX-FileCopyrightText: 2019-2020 Damien P. George
+#
+# SPDX-License-Identifier: MIT
+#
+# MicroPython uasyncio module
+# MIT license; Copyright (c) 2019-2020 Damien P. George
+"""
+Exceptions and classes related to asyncio Queue implementations.
+"""
+
+from . import event
```

Comment:
Hmm, looking at the other implementations - it looks like they don't re-use between the modules.  Should this use the underlying task queue to handle waiting for changes?

Reply:
I don't know what you mean about "don't reuse". But I have not studied the code in detail, and I'm not sure I would have an answer for you in any case.

Reply:
They pull only from `core` rather than from each other

Reply:
funcs.py for example doesn't have a need to look at event. stream.py (which is rather broken for Circuitpython) doesn't need to look at event. 
In your case, you do need to look at event. and not necessarily anything in core for the logic for this class.

I am curious though - It may be "better" to use the same "await _never" logic in core that event is based off of (as as as IOQueue in core)  Both from a standpoint of not having to share API with event (which seems helpful as a library) and from more control over your task execution than what event provides. not to mention event.py tries to load uio when importing it. 

Heres an example of using await _never() from core. This may also behave better as it gives you more control on what tasks to wake up - as it looks like set() in event will wake up every task wait()ing which may result in a few tasks going back to sleep due the queue being emptied by a prior task that was woken up. 

The implementation below only wakes up one task at a time to grab the available data. 

```python
class Queue:
    """
    A queue, useful for coordinating producer and consumer coroutines.

    If maxsize is less than or equal to zero, the queue size is infinite. If it
    is an integer greater than 0, then "await put()" will block when the
    queue reaches maxsize, until an item is removed by get().

    Unlike CPython's asyncio.Queue, this implementation is backed by a list rather
    than `collections.deque` because smaller boards may not have the library
    implemented.
    """

    def __init__(self, maxsize=2):
        self.maxsize = maxsize

        self._queue = []
        self._join_waiters = core.TaskQueue()
        self._join_counter = 0
        self._get_waiters = core.TaskQueue()
        self._put_waiters = core.TaskQueue()

    def _get(self):
        value = self._queue.pop(0)
        if self._put_waiters.peek():
            core._task_queue.push_head(self._put_waiters.pop_head())
        return value

    def _put(self, val):
        self._join_counter += 1
        self._queue.append(val)
        if self._get_waiters.peek():
            core._task_queue.push_head(self._get_waiters.pop_head())

    async def get(self):
        """
        Remove and return an item from the queue.

        If queue is empty, wait until an item is available.
        """
        while self.empty():
            self._get_waiters.push_head(core.cur_task)
            await core._never()
        return self._get()

    def get_nowait(self):
        """
        Remove and return an item from the queue.

        If queue is empty, raise QueueEmpty.
        """
        if self.empty():
            raise QueueEmpty()
        return self._get()

    async def put(self, val):
        """
        Put an item into the queue.

        If the queue is full, waits until a free
        slot is available before adding item.
        """
        while self.full():
            self._put_waiters.push_head(core.cur_task)
            await core._never()
        self._put(val)

    def put_nowait(self, val):
        """
        Put an item into the queue.

        If the queue is full, raises QueueFull.
        """
        if self.full():
            raise QueueFull()
        self._put(val)

    def qsize(self):
        """
        Number of items in this queue.
        """
        return len(self._queue)

    def empty(self):
        """
        Return True if the queue is empty.
        """
        return len(self._queue) == 0

    def full(self):
        """
        Return True if there are maxsize items in the queue.
        """
        return 0 < self.maxsize <= self.qsize()

    def task_done(self):
        """
        Indicate that a formerly enqueued task is complete.

        If a join() is currently blocking, it will resume when all items have
        been processed (meaning that a task_done() call was received for every
        item that had been put() into the queue).

        Raises ValueError if called more times than there were items placed in
        the queue.
        """
        if self._join_counter == 0:
            # Can't have less than 0
            raise ValueError("task_done() called too many times")

        self._join_counter -= 1

        if self._join_counter == 0:
            while self._join_waiters.peek():
                core._task_queue.push_head(self._join_waiters.pop_head())

    async def join(self):
        """
        Block until all items in the queue have been gotten and processed.
        """
        self._join_waiters.push_head(core.cur_task)
        await core._never()
```


example use:
```python
queue = asyncio.Queue()
counter = 0
get_counter = 0
hogged = False

async def add_to_queue():
    while True:
        global counter
        counter = counter + 1
        print("Putting " + str(counter))
        await queue.put("Hello number " + str(counter))
        await asyncio.sleep_ms(30)

async def read_from_queue():
    while True:
        global get_counter
        get_counter = get_counter + 1
        my_counter = get_counter
        print("Getting " + str(my_counter))
        print(str(await queue.get()))
        print("getting done " + str(my_counter))
        await asyncio.sleep_ms(500)
async def read_from_queue2():
    while True:
        global get_counter
        get_counter = get_counter + 1
        my_counter = get_counter
        print("Getting 2 " + str(my_counter))
        print(str(await queue.get()))
        print("getting 2 done" + str(my_counter))
        await asyncio.sleep_ms(1000)
        global hogged
        if my_counter > 50 and not hogged:
            hogged = True
            asyncio.create_task(hog_queue())
            asyncio.create_task(hog_queue())


async def hog_queue():
    print("Hogging")
    while get_counter < 200:
        global get_counter
        get_counter = get_counter + 1
        my_count = get_counter
        print("Getting Hog " + str(my_count))
        print(str(await queue.get()))
        print("getting Hog done" + str(my_count))

asyncio.create_task(add_to_queue())
asyncio.create_task(read_from_queue2())
asyncio.run(read_from_queue())
```

I only comment here as asyncio seems like it needs some love and I want to use cpy more and to me that means more async. Looking forward to helping work on this lib.

Reply:
funcs.py for example doesn't have a need to look at event. stream.py (which is rather broken for Circuitpython) doesn't need to look at event. 
In your case, you do need to look at event. and not necessarily anything in core for the logic for this class.

I am curious though - It may be "better" to use the same "await _never" logic in core that event is based off of (as as as IOQueue in core)  Both from a standpoint of not having to share API with event (which seems helpful as a library) and from more control over your task execution than what event provides. not to mention event.py tries to load uio when importing it. 

Heres an example of using await _never() from core. This may also behave better as it gives you more control on what tasks to wake up - as it looks like set() in event will wake up every task wait()ing which may result in a few tasks going back to sleep due the queue being emptied by a prior task that was woken up. 

The implementation below only wakes up one task at a time to grab the available data. 

```python
class Queue:
    """
    A queue, useful for coordinating producer and consumer coroutines.

    If maxsize is less than or equal to zero, the queue size is infinite. If it
    is an integer greater than 0, then "await put()" will block when the
    queue reaches maxsize, until an item is removed by get().

    Unlike CPython's asyncio.Queue, this implementation is backed by a list rather
    than `collections.deque` because smaller boards may not have the library
    implemented.
    """

    def __init__(self, maxsize=2):
        self.maxsize = maxsize

        self._queue = []
        self._join_waiters = core.TaskQueue()
        self._join_counter = 0
        self._get_waiters = core.TaskQueue()
        self._put_waiters = core.TaskQueue()

    def _get(self):
        value = self._queue.pop(0)
        if self._put_waiters.peek():
            core._task_queue.push_head(self._put_waiters.pop_head())
        return value

    def _put(self, val):
        self._join_counter += 1
        self._queue.append(val)
        if self._get_waiters.peek():
            core._task_queue.push_head(self._get_waiters.pop_head())

    async def get(self):
        """
        Remove and return an item from the queue.

        If queue is empty, wait until an item is available.
        """
        while self.empty():
            self._get_waiters.push_head(core.cur_task)
            await core._never()
        return self._get()

    def get_nowait(self):
        """
        Remove and return an item from the queue.

        If queue is empty, raise QueueEmpty.
        """
        if self.empty():
            raise QueueEmpty()
        return self._get()

    async def put(self, val):
        """
        Put an item into the queue.

        If the queue is full, waits until a free
        slot is available before adding item.
        """
        while self.full():
            self._put_waiters.push_head(core.cur_task)
            await core._never()
        self._put(val)

    def put_nowait(self, val):
        """
        Put an item into the queue.

        If the queue is full, raises QueueFull.
        """
        if self.full():
            raise QueueFull()
        self._put(val)

    def qsize(self):
        """
        Number of items in this queue.
        """
        return len(self._queue)

    def empty(self):
        """
        Return True if the queue is empty.
        """
        return len(self._queue) == 0

    def full(self):
        """
        Return True if there are maxsize items in the queue.
        """
        return 0 < self.maxsize <= self.qsize()

    def task_done(self):
        """
        Indicate that a formerly enqueued task is complete.

        If a join() is currently blocking, it will resume when all items have
        been processed (meaning that a task_done() call was received for every
        item that had been put() into the queue).

        Raises ValueError if called more times than there were items placed in
        the queue.
        """
        if self._join_counter == 0:
            # Can't have less than 0
            raise ValueError("task_done() called too many times")

        self._join_counter -= 1

        if self._join_counter == 0:
            while self._join_waiters.peek():
                core._task_queue.push_head(self._join_waiters.pop_head())

    async def join(self):
        """
        Block until all items in the queue have been gotten and processed.
        """
        self._join_waiters.push_head(core.cur_task)
        await core._never()
```


example use:
```python
queue = asyncio.Queue()
counter = 0
get_counter = 0
hogged = False

async def add_to_queue():
    while True:
        global counter
        counter = counter + 1
        print("Putting " + str(counter))
        await queue.put("Hello number " + str(counter))
        await asyncio.sleep_ms(30)

async def read_from_queue():
    while True:
        global get_counter
        get_counter = get_counter + 1
        my_counter = get_counter
        print("Getting " + str(my_counter))
        print(str(await queue.get()))
        print("getting done " + str(my_counter))
        await asyncio.sleep_ms(500)
async def read_from_queue2():
    while True:
        global get_counter
        get_counter = get_counter + 1
        my_counter = get_counter
        print("Getting 2 " + str(my_counter))
        print(str(await queue.get()))
        print("getting 2 done" + str(my_counter))
        await asyncio.sleep_ms(1000)
        global hogged
        if my_counter > 50 and not hogged:
            hogged = True
            asyncio.create_task(hog_queue())
            asyncio.create_task(hog_queue())


async def hog_queue():
    print("Hogging")
    while get_counter < 200:
        global get_counter
        get_counter = get_counter + 1
        my_count = get_counter
        print("Getting Hog " + str(my_count))
        print(str(await queue.get()))
        print("getting Hog done" + str(my_count))

asyncio.create_task(add_to_queue())
asyncio.create_task(read_from_queue2())
asyncio.run(read_from_queue())
```

I only comment here as asyncio seems like it needs some love and I want to use cpy more and to me that means more async. Looking forward to helping work on this lib.

Reply:
Refactored a bit to use this.

I took a bit of a detour trying to get tests working.  Because this is using circuitpython async/await the tests just blow up when we try to run using python 3.

However, I was able to verify this as working by including it in the `lib` of an rp2040 board as `py` files